### PR TITLE
Lg/candid rev pre split

### PR DESCRIPTION
**Pull request by lsgunnlsgunn**
This is a rough cut at moving the Candid user's guide draft into the revamped main SDK site navigation.
My thought is to split it into smaller sections, add links, make it somewhat more consistent in tone with other sections, etc.

I thought it would be easier to see the proposed changes if I made them in the index.adoc file (maybe it would have been simpler to view the smaller files but you wouldn't see diffs that way). Most of the changes are pretty minor and I didn't really touch the reference sections, but I did try to add some headings to a few topics and sometimes to interpret where a topic was going ... I figure it might need a little more wrangling.

Here's the big picture:
image

### CODE REVIEW

File: docs/modules/candid-guide/index.adoc [hunk 1]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
-
-* JSON, XML etc. are merely data description languages. They describe how to map the values that we want to transfer to bytes or characters, but they do not describe the services as a whole.
-
-* Candid implementations map the Candid value directly to types and values of the host language; in Candid, programmers do not construct or deconstruct some abstract `Candid` value.
+At first glance, you might think that other technologies, such as JSON, XML, or Protobuf, would suffice. 
+However, Candid provides a unique combination of features that are not found in these other technologies.
+The features that make Candid particularly well-suited for developing applications for the Internet Computer include the following:
 
-* Candid defines what is a “safe upgrade” in a sound and compositional way.
+* Many interface description languages like JSON, XML, and Protobuf describe how to map individual values to bytes or characters, but they do not describe services as a whole. These languages focus on the data types you want to transfer instead of the methods that make use of those data types.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:48]: The gist is that they aren't interface description languages (and not normally called that). They are merely data description languages. So I would reword that slightly.

File: docs/modules/candid-guide/index.adoc [hunk 2]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 
-* Candid is inherently higher order: You can pass not just plain data, but also references to services and methods, which is an important feature on the Internet Computer. Furthermore, Candid’s notion of safe upgrades takes such higher-order use into account.
+* Candid implementations map the Candid value directly to types and values of the host language. In Candid, programmers do not construct or deconstruct some abstract `+Candid+` value.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:50]:
```suggestion
* Candid implementations map the Candid value directly to types and values of the host language. With Candid, programmers do not construct or deconstruct some abstract `+Candid+` value.
```

File: docs/modules/candid-guide/index.adoc [hunk 3]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 
-* Candid has built-in support for Internet-Computer specific features, such as the `query` annotation.
+* Candid defines rules for how software can be upgraded in a sound and compositional way.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:52]:
```suggestion
* Candid defines rules for how services and their interface can be upgraded in a sound and compositional way.
```

File: docs/modules/candid-guide/index.adoc [hunk 4]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 
-== Concepts and workflows
+* Candid is inherently a higher order languages. With Candid, you can pass more than plain data, including references to services and methods. The Candid support for safe upgrades takes such higher-order use into account.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:54]:
```suggestion
* Candid is inherently a higher-order languages. With Candid, you can pass more than plain data, including references to services and methods. Candid's support for safe upgrades takes such higher-order use into account.
```

File: docs/modules/candid-guide/index.adoc [hunk 5]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 
-Furthermore, the set of types is chosen to allow a natural mapping to the respective host language (Motoko, Rust, JavaScript, etc.), settling on reasonable, canonical choices.
+Candid supports this set of types to allow a natural mapping of data types based on reasonable, canonical choices suitable for each host language, whether you are writing your code in Motoko, Rust, JavaScript, or some other language.
 
 [#candid-service-descriptions]
-=== Candid service descriptions
+== Candid service descriptions
 
-With these Candid types, we now can describe a service. Such a service description file can be written by hand, but typically it is generated from a service implementation, as described later.
+Once you are familiar with the Candid types, you can use them describe a service. A Candid service description file—a `+.DID+` file—can either be written by hand or generated from a service implementation.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:80]:
```suggestion
Once you are familiar with the Candid types, you can use them to describe a service. A Candid service description file—a `+.DID+` file—can either be written by hand or generated from a service implementation.
```

File: docs/modules/candid-guide/index.adoc [hunk 6]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 service : {
   ping : () -> ();
 }
 ....
 
-This says that the service supports a public method called `ping`. Method names can be arbitrary strings, and you can quote them (`"method with spaces"`) if they are not plain identifiers.
+This example describes a service supports a single public method called `+ping+`. 
+Method names can be arbitrary strings, and you can quote them (`"method with spaces"`) if they are not plain identifiers.
+
+Methods declare a _sequence_ of arguments and result types. In the case of this `+ping+` method, no arguments are passed and none are returned, so the empty sequence `+()+ -> ()+` is used.
```
Comment (nomeata) [inline docs/modules/candid-guide/index.adoc:101]:
```suggestion
Methods declare a _sequence_ of arguments and result types. In the case of this `+ping+` method, no arguments are passed and none are returned, so for both the empty sequence `+()+` is used. 
```

File: docs/modules/candid-guide/index.adoc [hunk 7]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
+
+This service description is not very useful, so let's add a simple `+ping+` method:
 ....
 service : {
   ping : () -> ();
 }
 ....
 
-This says that the service supports a public method called `ping`. Method names can be arbitrary strings, and you can quote them (`"method with spaces"`) if they are not plain identifiers.
+This example describes a service supports a single public method called `+ping+`.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:98]:
```suggestion
This example describes a service that supports a single public method called `+ping+`. 
```

File: docs/modules/candid-guide/index.adoc [hunk 8]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
+
+The simplest service description file that describes a service with no public methods would look like this:
 
-The simplest service description is
 ....
 service : {}
 ....
-and describes a service with no public methods. This is not very useful, so let's add a simple method:
+
+This service description is not very useful, so let's add a simple `+ping+` method:
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:91]:
```suggestion
This service is not very useful, so let's add a simple `+ping+` method:
```

File: docs/modules/candid-guide/index.adoc [hunk 9]
```diff
@@ -127,26 +148,40 @@ service address_book : {
...
+For example:
+....
+service address_book : {
+  set_address: (name : text, addr : address) -> ();
+  get_address: (name : text) -> (opt address) query;
+}
+....
+
+This annotation indicates that the `+get_address+` method can be invoked as an {IC} *query call*.
+As discussed in link:developers-guide:concepts/canisters-code{outfilesuffix}#query-update[Query and update methods], a query provides an efficient way to retrieve information from a canister without going through consensus so being able to identify a method as a query one of the key benefits of using Candid to interact with the {IC}.
```
Comment (nomeata) [inline docs/modules/candid-guide/index.adoc:163]: Thanks for digging out the link!

File: docs/modules/candid-guide/index.adoc [hunk 10]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
+Candid supports this set of types to allow a natural mapping of data types based on reasonable, canonical choices suitable for each host language, whether you are writing your code in Motoko, Rust, JavaScript, or some other language.
 
 [#candid-service-descriptions]
-=== Candid service descriptions
+== Candid service descriptions
 
-With these Candid types, we now can describe a service. Such a service description file can be written by hand, but typically it is generated from a service implementation, as described later.
+Once you are familiar with the Candid types, you can use them describe a service. A Candid service description file—a `+.DID+` file—can either be written by hand or generated from a service implementation.
+
+Before you explore how to generate a service description file for a specific host language, let's take a closer look at the structure of the file and its constituent parts.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:82]: I'd avoid most references to "files", since they are just a way in which such a description can be physically stored, but on the IC they may occur in other form as well.
```suggestion
Before you explore how to generate a service descriptions for a specific host language, let's take a closer look at their structure and its constituent parts.
```

File: docs/modules/candid-guide/index.adoc [hunk 11]
```diff
@@ -289,23 +320,56 @@ If you want to create a new Candid implementation, this document will not suffic
...
-TIP: Subtypes are the types you can change your method _results_ to, and supertypes are the types that you can change your method _arguments_ to.
+Subtypes are the types you can change your method _results_ to.
+Supertypes are the types that you can change your method _arguments_ to.
 
-We only list the particular subtypes and supertypes in each section, and do not repeat every time that that the type `empty` is a subtype of any type, and that the types `reserved` and `opt t` are supertypes of any types. See the sections on these types for more details.
+You should note that this reference only lists the specific subtypes and supertypes that are relevant for each type. 
+It does not repeat common information about subtypes and supertypes that can apply to any type. 
+For example, the reference does not list `+empty+` as a subtype because it can be a subtype of any other type.
+Similarly, the types `+reserved+` and `+opt t+` are not listed as supertypes of specific types because they are supertypes of any type. 
+The `+empty+`, `+reserved+`, and `+opt t+` types are discussed separately, independent of other types.
```
Comment (nomeata) [inline docs/modules/candid-guide/index.adoc:363]:
```suggestion
See the sections about the <<type-empty,`+empty+` type>>, the <<type-reserved,`+reserved+` type>> and the <<type-opt,`+opt t+` type>> for details about their subtyping rules.
```

**Review by nomeata — APPROVED** [review]
Yay

File: docs/modules/candid-guide/index.adoc [hunk 12]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
-=== Candid service descriptions
+== Candid service descriptions
 
-With these Candid types, we now can describe a service. Such a service description file can be written by hand, but typically it is generated from a service implementation, as described later.
+Once you are familiar with the Candid types, you can use them describe a service. A Candid service description file—a `+.DID+` file—can either be written by hand or generated from a service implementation.
+
+Before you explore how to generate a service description file for a specific host language, let's take a closer look at the structure of the file and its constituent parts.
+but typically it is , as described later.
+
+The simplest service description file that describes a service with no public methods would look like this:
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:85]:
```suggestion
The simplest service description specifies a service with no public methods and would look like this:
```

File: docs/modules/candid-guide/index.adoc (hunk 6 above)
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:101]:
```suggestion
Methods declare a _sequence_ of arguments and result types. In the case of this `+ping+` method, no arguments are passed and no results are returned, so the empty sequence `+()+` is used for both arguments and results. 
```

File: docs/modules/candid-guide/index.adoc [hunk 13]
```diff
@@ -23,84 +27,99 @@ service counter : {
...
 
-The method `divMod` expects and returns two values, all of type `int`. In its signature, we use the ability to name the parameters and result values.
+The method `+divMod+` expects and returns two values, all of type `+nat+`.
 
-[CAUTION]
-====
-Naming the parameters or results of a method is purely for documentation purposes, but does not actually change the method’s type, or the values being passed. Parameter and results are identified by their _position_, independent of the name.
+== Naming arguments and results
+
+In this example service description file, the signature for the `+divMod+` method includes names for the argument and result values.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:119]:
```suggestion
In the previous example, the signature for the `+divMod+` method includes names for the argument and result values. 
```

File: docs/modules/candid-guide/index.adoc [hunk 14]
```diff
@@ -127,26 +148,40 @@ service address_book : {
 
 These type definitions merely abbreviate an _existing_ type, they do not define a new type. It does not matter whether you use `address` in the function signature, or write out the records. Also, two abbreviations with different names but equivalent definitions, describe the same type and are interchangeable. In other words, Candid uses _structural_ typing.
 
-In the last example, we also see the use of the `query` annotation. This Internet-Computer specific annotation indicates that the method `get_address` can be invoked using a query call (TODO: reference general documentation about query calls here)
+== Specifying a query method
 
-=== Encoding and decoding
+In the last example service description file, you might have noticed the use of the `+query+` annotation for the `+get_address+` method.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:153]:
```suggestion
In the last example, you might have noticed the use of the `+query+` annotation for the `+get_address+` method. 
```

File: docs/modules/candid-guide/index.adoc (hunk 9 above)
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:163]:
```suggestion
As discussed in link:developers-guide:concepts/canisters-code{outfilesuffix}#query-update[Query and update methods], a query provides an efficient way to retrieve information from a canister without going through consensus, so being able to identify a method as a query one of the key benefits of using Candid to interact with the {IC}.
```

File: docs/modules/candid-guide/index.adoc [hunk 15]
```diff
@@ -127,26 +148,40 @@ service address_book : {
...
 
 * The Candid binary format always encodes _sequences_ of values, because methods parameters and results are sequences of types.
 
-* The binary format is quite compact. A `(vec nat64)` with 125 000 entries takes 1 000 007 bytes.
+* The binary format is quite compact. A `+(vec nat64)+` with 125 000 entries takes 1 000 007 bytes.
 
 * The binary is self-describing, and includes a (condensed) description of type of the values therein. This allows the receiving side to detect if a message was sent at a different, incompatible type.
 
-* As long as the sending side serializes the arguments at the same type that the receiving side expects them, deserialization will succeed.
+* As long as the sending-side serializes the arguments at the same type that the receiving side expects them, deserialization will succeed.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:181]:
```suggestion
* As long as the sender serializes the arguments at the same type that the receiver expects them, deserialization will succeed.
```

File: docs/modules/candid-guide/index.adoc [hunk 16]
```diff
@@ -228,34 +271,22 @@ actor {
   public func subscribe(handler : func (Int) -> async ()) { … }
 }
 ....
-and the compiler can automatically generate the interface shown above.
+
+When you compile this program, the {proglang} compiler automatically generates the Candid service description file with the interface shown above.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:275]:
```suggestion
When you compile this program, the {proglang} compiler automatically generates a Candid service description file with the interface shown above.
```

File: docs/modules/candid-guide/index.adoc [hunk 17]
```diff
@@ -289,23 +320,56 @@ If you want to create a new Candid implementation, this document will not suffic
 
 Additionally, the official https://github.com/dfinity/candid/tree/master/test[Candid test data] is useful to make sure your implementation is compatible with Candid, even in slightly more obscure corner cases.
 
-[#reference]
 == Reference
 
-This section lists all the types supported by Candid, including their type syntax, the syntax for the textual representation, their upgrading rules and the corresponding types in Rust, Motoko and Javascript.
+This section provides links to Candid tools and libraries for currently supported host languages and detailed reference information about Candid supported types.
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:325]: Hm, it feels odd to lump the description of tools under the Candid language reference section. I'd move it somewhere else.

File: docs/modules/candid-guide/index.adoc [hunk 18]
```diff
@@ -289,23 +320,56 @@ If you want to create a new Candid implementation, this document will not suffic
...
+=== Supported types
+
+This section lists all the types supported by Candid.
+For each type, the reference includes the following information:
+
+* Type syntax and the syntax for the textual representation of the type.
+* Upgrade rules for the type.
+* Corresponding types in Rust, Motoko and Javascript.
+
+== Subtypes and supertypes
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:354]: Starting a new section here is a bit of an odd break (and it should probably be a subsection).
```suggestion
Upgrade rules are given in terms of the possible _subtypes_ and _supertypes_ of a type.
```

File: docs/modules/candid-guide/index.adoc [hunk 19]
```diff
@@ -289,23 +320,56 @@ If you want to create a new Candid implementation, this document will not suffic
...
-We only list the particular subtypes and supertypes in each section, and do not repeat every time that that the type `empty` is a subtype of any type, and that the types `reserved` and `opt t` are supertypes of any types. See the sections on these types for more details.
+You should note that this reference only lists the specific subtypes and supertypes that are relevant for each type. 
+It does not repeat common information about subtypes and supertypes that can apply to any type. 
+For example, the reference does not list `+empty+` as a subtype because it can be a subtype of any other type.
+Similarly, the types `+reserved+` and `+opt t+` are not listed as supertypes of specific types because they are supertypes of any type. 
+The `+empty+`, `+reserved+`, and `+opt t+` types are discussed separately, independent of other types.
 
 [#type-text]
-=== Type text
+== Type text
```
Comment (rossberg) [inline docs/modules/candid-guide/index.adoc:366]: This and all the following sections are part of the language reference, so should probably remain to be subsections of Reference.

File: docs/modules/candid-guide/index.adoc (hunk 11 above)
Comment (lsgunnlsgunn) [reply docs/modules/candid-guide/index.adoc]: Modified

**Review by lsgunnlsgunn — COMMENTED** [review]
no changes to content from this point.
